compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -2328,21 +2328,78 @@ public void visit(BLangXMLNS xmlnsNode) {
     }
 
     public void visit(BLangCompoundAssignment compoundAssignment) {
-        BLangAssignment assignStmt = (BLangAssignment) TreeBuilder.createAssignmentNode();
-        assignStmt.pos = compoundAssignment.pos;
-        BLangVariableReference varRef = compoundAssignment.varRef;
+        // If compound Assignment is an index based expression such as a[f(1, foo)][3][2] += y,
+        // should return a block statement which is equivalent to
+        // var $temp1$ = 2;
+        // var $temp2$ = 3;
+        // var $temp3$ = a[f(1, foo)];
+        // a[$temp3$][$temp2$][$temp1$] = a[$temp3$][$temp2$][$temp1$] + y;
+        if (compoundAssignment.varRef.getKind() == NodeKind.INDEX_BASED_ACCESS_EXPR) {
+            int indexExprNumber = 0;
+            List<BLangStatement> statements = new ArrayList<>();
+            List<BLangSimpleVarRef> varRef = new ArrayList<>();
+            List<BType> type = new ArrayList<>();
+
+            BLangIndexBasedAccess indexBasedAccess = (BLangIndexBasedAccess) compoundAssignment.varRef;
+            // Extract the index Expressions from compound assignment and create variable definitions. ex:
+            // var $temp1$ = 2;
+            // var $temp2$ = 3;
+            // var $temp3$ = a[f(1, foo)];
+            do {
+                indexExprNumber += 1;
```

Comment:
Even though this works atm, shouldn't this be a global variable instead? We might be reusing the same variable names with compound assignments if there are multiple being rewritten within the same block.

Reply:
Fixed.